Documentation/Contributing/UnitTests.md

```diff
@@ -0,0 +1,133 @@
+
+# Writing tests for your code
+To ensure MRTK being a stable and reliable toolkit, every feature should come with unit tests and some application in one of the example scenes.
+
+Preferrably when fixing a bug there should also be a test added to avoid running into the same issue again in the future.
+
+Having good test coverage in a big codebase like MRTK is crucial for stability and having confidence when doing changes in code.
+
+
+MRTK uses the [Unity Test Runner](https://docs.unity3d.com/Manual/testing-editortestsrunner.html) which uses a Unity integration of [NUnit](https://nunit.org/). 
+
+This guide will give a starting point on how to add tests to MRTK. It will not explain the [Unity Test Runner](https://docs.unity3d.com/Manual/testing-editortestsrunner.html) and [NUnit](https://nunit.org/) which can be looked up in the links provided.
+
+There's two types of tests that can be added for new code
+
+* Edit mode tests
+* Play mode tests
+
+## Play mode tests
+
+Play mode tests will be executed in Unity's play mode and should be added into MixedRealityToolkit.Tests > PlaymodeTests. 
+To create a new test the following template can be used: (todo: Setup() and teardown should be moved into a base class of mrtk tests)
+
+``` csharp
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#if !WINDOWS_UWP
+// When the .NET scripting backend is enabled and C# projects are built
+// Unity doesn't include the the required assemblies (i.e. the ones below).
+// Given that the .NET backend is deprecated by Unity at this point it's we have
+// to work around this on our end.
+using NUnit.Framework;
+using UnityEngine.TestTools;
+using System.Collections;
+
+namespace Microsoft.MixedReality.Toolkit.Tests
+{
+    public class ExampleTest : IPrebuildSetup
+    {
+        public void Setup()
+        {
+            PlayModeTestUtilities.EnsureTextMeshProEssentials();
+        }
+
+
+        [TearDown]
```

Comment:
I think that TearDown is a nunit concept right - it feels a bit weird seeing a nunit teardown but a unity setup (i.e. Setup doesn't have a corresponding SetUp attribute, I guess cause it doesn't need it)

Reply:
agree it's a bit weird - @googlan probably knows why it's done that way

Reply:
It looks a bit weird, but these are completely different things in this case. Most tests use NUnit `[Setup]` and `[TearDown]` and don't inherit from `IPrebuildStep`. I think having an empty `[Setup]` method (or calling `TestUtilities.InitializeMixedRealityToolkit(true);` in it) in this example with a comment would help.

Basically `Setup` comes from inheriting from `IPrebuildSetup` and is called before entering play mode (see comment above). This one is Unity-specific extension on top of NUnit.
Method, marked with NUnit `[Setup]` is called before the test, but during the play mode (same with `TearDown`). Both come directly from NUnit runner.

Reply:
> Thanks for making this! You should check out Lars's PR and update the setup/teardown methods, but otherwise looks good.

thanks Julia! had a look at the PR - i've adjusted the template - now i'm doing the init and destroy of mrtk in the setup and teardown methods as well. i think it makes sense to add lars new utility methods there as well after the PR has been merged. - I would also like to add some more guides / snippets in future for creating certain ux elements - so this is not the final guide yet, just a starting point so there's something available for ppl that haven;t written any tests yet.